code_generator/GdkCodeGenerator/Templates/UnityGameObjectComponentDispatcherGenerator.tt

```diff
@@ -60,21 +62,35 @@ namespace <#= qualifiedNamespace #>
                 return <#= unityComponentDefinition.Id #>;
             }
 
-            public override void InvokeOnComponentUpdateUserCallbacks(GameObjectDispatcherSystem gameObjectDispatcherSystem)
+            public override void InvokeOnAuthorityChangeLifecycleCallbacks(Dictionary<int, MonoBehaviourActivationManager> entityIndexToManagers)
+            {
+                var authoritiesChangedTags = AuthoritiesChangedComponentGroup.GetComponentArray<AuthoritiesChanged<<#= componentDetails.TypeName #>>>();
+                var entities = AuthoritiesChangedComponentGroup.GetEntityArray();
+                for (var i = 0; i < entities.Length; i++)
+                {
+                    var activationManager = entityIndexToManagers[entities[i].Index];
+                    for (var j = 0; j < authoritiesChangedTags[i].Buffer.Count; j++)
+                    {
+                        activationManager.ChangeAuthority(<#= unityComponentDefinition.Id #>, authoritiesChangedTags[i].Buffer[j]);
+                    }
+                }
+            }
+
+            public override void InvokeOnComponentUpdateUserCallbacks(Dictionary<int, ReaderWriterStore> readerWriterStores)
             {
 <# if (fieldDetailsList.Count > 0) { #>
                 var entities = ComponentsUpdatedComponentGroup.GetEntityArray();
                 var updateLists = ComponentsUpdatedComponentGroup.GetComponentArray<ComponentsUpdated<<#= componentDetails.TypeName #>.Update>>();
                 for (var i = 0; i < entities.Length; i++)
                 {
-                    var manager = gameObjectDispatcherSystem.GetSpatialOSBehaviourManager(entities[i].Index);
-                    if (!manager.TryGetReadersWriters(<#= unityComponentDefinition.Id #>, out var readers))
+                    var readerWriterStore = readerWriterStores[entities[i].Index];
+                    if (!readerWriterStore.TryGetReaderWritersForComponent(<#= unityComponentDefinition.Id #>, out var readers))
                     {
                         continue;
                     }
 
                     var updateList = updateLists[i];
-                    foreach (<#= componentDetails.ComponentName #>.ReaderWriterImpl reader in readers)
+                    foreach (var reader in readers.OfType<<#= componentDetails.ComponentName #>.ReaderWriterImpl>())
```

Comment:
Do you mean `Cast`? https://stackoverflow.com/questions/4015930/when-to-use-cast-and-oftype-in-linq

Reply:
In which case please use regular explicit cast. Cast will return an IEnumerable which will allocate on foreach. While probably not a huge problem it is unnecessary.